File: 0099-recover-binary-search-tree/0099-recover-binary-search-tree.py
from typing import Optional


# Definition for a binary tree node.
class TreeNode:

  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:

  def recoverTree(self, root: Optional[TreeNode]) -> None:
    """Do not return anything, modify root in-place instead."""
    first = None
    second = None
    prev = None

    def inorder(node: Optional[TreeNode]):
      nonlocal first, second, prev
      if not node:
        return

      inorder(node.left)

      # An inversion occurs when prev node value >= current node value
      if prev and prev.val > node.val:
        if not first:
          first = prev
        second = node

      prev = node
      inorder(node.right)

    inorder(root)

    # Swap the values of the two identified nodes
    if first and second:
      first.val, second.val = second.val, first.val

File: 0099-recover-binary-search-tree/0099-recover-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    TreeNode* first = nullptr;
    TreeNode* second = nullptr;
    TreeNode* prev = nullptr;

    void inorder(TreeNode* node) {
        if (!node) return;

        inorder(node->left);

        // Detect inversion
        if (prev && prev->val > node->val) {
            if (!first) {
                first = prev; // First violation: select the larger node
            }
            second = node;    // Both violations: update to the smaller node
        }
        prev = node;

        inorder(node->right);
    }

public:
    void recoverTree(TreeNode* root) {
        inorder(root);

        // Swap the values back to restore BST properties
        if (first && second) {
            std::swap(first->val, second->val);
        }
    }
};